File: solutions/c/sieve/1/sieve.c
#include "sieve.h"
#include <stdbool.h>
#include <string.h>
#include <stdio.h>

uint32_t sieve(uint32_t limit, uint32_t *primes, size_t max_primes) {
    bool sieve[max_primes + 1];
    memset(sieve, true, sizeof(sieve));
    for (uint32_t p=2; p*p <= max_primes; p++) {
        if (sieve[p])
            for (uint32_t i = p*p; i <= limit; i+=p)
                sieve[i] = false;
    }
    uint32_t count = 0;
    for (uint32_t i = 2; i <= limit; i++)
        if (sieve[i])
            primes[count++] = i;
    return count;
}

File: solutions/python/sieve/1/sieve.py
from itertools import takewhile


def primes(limit):
    sieve = {k: True for k in range(2, 1003)}
    for i, p in sieve.items():
        if p:
            for j in range(i * i, len(sieve), i):
                sieve[j] = False
    primes = list(filter(lambda k: sieve[k], sieve))
    return list(takewhile(lambda n: n <= limit, primes))
